### PR TITLE
rgw:  update op_mask of user via admin rest api

### PR DESCRIPTION
**Pull request by mslovy**

### CODE REVIEW

**Comment by mslovy** [comment]
any take a look?

File: src/rgw/rgw_rest_user.cc [hunk 1]
```diff
@@ -268,6 +270,17 @@ void RGWOp_User_Modify::execute()
   if (s->info.args.exists("system"))
     op_state.set_system(system);
 
+  if (!op_mask_str.empty() && s->info.args.exists("op-mask")) {
+    uint32_t op_mask;
+    int ret = rgw_parse_op_type_list(op_mask_str, &op_mask);
+    if (ret < 0)
+    {
+      ldout(s->cct, 0) << "failed to parse op_mask: " << op_mask_str << dendl;
+      return;
```
[the remark this answers is not in the document]
Comment (mslovy) [reply src/rgw/rgw_rest_user.cc]: no，I don't think so. I find  rgw_parse_op_type_list always reutrn 0 now, shoud we still need to preserve the ret.
this comes from:  rgw_admin.cc,  I keep it both the same. I'm not quite clear why should we need the return value here.

**Comment by mslovy** [comment]
@yehudasa should we need to consider make the user no any permission to do anything, which means NULL below.  And currently, it seems not support such kinds of operation.

{ {"*",  RGW_OP_TYPE_ALL},
                  {"read",  RGW_OP_TYPE_READ},
		  {"write", RGW_OP_TYPE_WRITE},
		  {"delete", RGW_OP_TYPE_DELETE},
		  {NULL, 0} };

**Comment by yehudasa** [comment]
@mslovy the NULL option marks the end of the array. You can add a new line with "none" string.

**Comment by mslovy** [comment]
still problems like ,  radosgw-admin user modify --op-mask "none",   "blabla",   "aha" .   any strings that not "read", "write", "delete" will lead to  op-mask become 0.  it is fine for us because our dashboard prevent user to enter the invalid params like "blabla".  Do you think any checks here to prevent invalid params?

**Comment by yehudasa** [comment]
@mslovy we shouldn't silently ignore parsing errors there. If we don't recognize the value we should return error.

**Comment by mslovy** [comment]
@tchaikov , cmake switch to c++17 and something goes wrong?

**Comment by tchaikov** [comment]
it's a networking issue. should be addressed by https://github.com/ceph/ceph-build/pull/1076

retest this please.

**Comment by mslovy** [comment]
retest this please

**Comment by tchaikov** [comment]
@mslovy might want to fix the broken tests.

**Comment by mslovy** [comment]
retest this please

**Comment by mslovy** [comment]
retest this please

**Comment by mslovy** [comment]
@yehudasa , can you take a look again, any other comments?

**Comment by mslovy** [comment]
anyone review this? @tchaikov

**Comment by mslovy** [comment]
hi, guys.  anyone think it is meaningful?

**Comment by tchaikov** [comment]
@mslovy i am not qualified to review rgw changes. probably you could ping @yehudasa on this?

**Comment by mslovy** [comment]
ping, @yehudasa

**Comment by mslovy** [comment]
@cbodley   Currentlly , we cannot update the op_mask for a user for admin rest api, and so that it is hard to intergrate with the dashboard?

**Comment by cbodley** [comment]
thanks @mslovy. will this need a backport to nautilus?

**Comment by mslovy** [comment]
yeah, I prefer to backport to nautilus

**Comment by cbodley** [comment]
@mslovy ok, i created http://tracker.ceph.com/issues/39084 to track the backport - can you please add this line to your commit message?
> Fixes: http://tracker.ceph.com/issues/39084

**Comment by mslovy** [comment]
okey,  updated

**Comment by cbodley** [comment]
http://pulpito.ceph.com/cbodley-2019-04-23_20:35:16-rgw-wip-cbodley-testing-distro-basic-smithi/

lots of valgrind issues from http://tracker.ceph.com/issues/39449

**Pull request merged**